File: graph_clone.py
# Definition for a undirected graph node
class UndirectedGraphNode:
    def __init__(self, x):
        self.label = x
        self.neighbors = []

class Solution:
    # @param node, a undirected graph node
    # @return a undirected graph node
    def cloneGraph(self, node):
        if node is None:
            return None

        new_nodes = {}
        parent = {node : None}
        self.dfs(node, new_nodes, parent)

        for k in parent.keys():
            for n in k.neighbors:
                new_nodes[k.label].neighbors.append(new_nodes[n.label])

        return new_nodes[node.label]

    def dfs(self, node, new_nodes, parent):
        new_nodes[node.label] = UndirectedGraphNode(node.label)

        for n in node.neighbors:
            if n not in parent:
                parent[n] = node
                self.dfs(n, new_nodes, parent)

File: C++_version/graph_clone.cpp
#include <vector>
#include <map>
#include <unordered_map>
#include <iostream>

using namespace std;


  // Definition for undirected graph.
 struct UndirectedGraphNode {
      int label;
      vector<UndirectedGraphNode *> neighbors;
      UndirectedGraphNode(int x) : label(x) {};
 };
 


// struct KeyHasher
// {
//   std::size_t operator()(const Key& k) const
//   {
//     using std::size_t;
//     using std::hash;
//     using std::string;

//     return ((hash<string>()(k.first)
//              ^ (hash<string>()(k.second) << 1)) >> 1)
//              ^ (hash<int>()(k.third) << 1);
//   }
// };

class Solution {
public:
    UndirectedGraphNode *cloneGraph(UndirectedGraphNode *node) {
        if (node == nullptr)
        	return nullptr;

        unordered_map<int, UndirectedGraphNode*> new_nodes;
        unordered_map<UndirectedGraphNode*, UndirectedGraphNode*> parent;

        parent[node] = nullptr;
        dfs(node, new_nodes, parent);

        for (auto item : parent) {
        	auto k = item.first;
        	for (auto n : k->neighbors) {
        		new_nodes[k->label]->neighbors.push_back(new_nodes[n->label]);
        	}
        }

        return new_nodes[node->label]; 
    }

    void dfs(UndirectedGraphNode* node, unordered_map<int, UndirectedGraphNode*>& new_nodes, unordered_map<UndirectedGraphNode*, UndirectedGraphNode*>& parent) {
    	new_nodes[node->label] = new UndirectedGraphNode(node->label);

    	for (auto n : node->neighbors) {
    		if (parent.find(n) == parent.end()) {
    			parent[n] = node;
    			dfs(n, new_nodes, parent);
    		}
    	}
    }
};

int main()
{
	Solution sol;

	return 0;
}